format/nexus.py

```diff
@@ -105,7 +105,7 @@ def find_entries(nx_file, entry):
 
     def visitor(name, obj):
         if "NX_class" in obj.attrs:
-            if obj.attrs["NX_class"] in [
+            if numpy.string_(obj.attrs["NX_class"]) in [
```

Comment:
et seq. 

while this _works_ for me i.e. does not change anything I am slightly confused why the strings need to be cast here. At the moment for me the comparisons work fine in Python 2 or 3 with files we write at Diamond - which I am not saying are right, but they are a reference point. If the string is a string then the string is a string, so why does it need to be cast to a numpy string here?

Humbly opine: there is something in how you are making the files themselves which is where the underlying error is, and this is simply patching over the cracks left behind. Maybe the better thing to do would be look at that code? 🤔

I'm not hostile to these changes, I just want to understand why they are necessary.

Reply:
So I don't know why the files you write in DLS are different than the files I write for SwissFEL. Are you using h5py or c++ directly?

In py3 h5py when I create a string array (see [example](https://github.com/cctbx/cctbx_project/blob/master/xfel/swissfel/jf16m_cxigeom2nexus.py#L126-L127)) I have to use numpy.string_, which resolves to byte string.  Otherwise the h5py function create_dataset fails.  But the underlying string is written in unicode.  h5dump confirms this.  Therefore, the line you indicate fails, unless it is first cast back to byte string.

Reply:
We use some weird local thing which I think is Java bindings to the HDF5 C API to do the file writing... I expect we could pull this out of the context of Eiger / Jungfrau / whatever and just try making an HDF5 file full of strings (even a couple would do it) actually work. 

Am happy to help with that - I would rather understand what is going on here. Feels to me exactly like a string encoding thing - I wonder if before write the strings should be ascii encoded explicitly?

Reply:
Ok, if it's the C++ API then I could believe you could write the strings as ascii encoded, which is how py2 works, but for h5py in py3, none of the tricks I tried worked.  I always got UTF8 encoded strings.

Reply:
We should be able to fix that - as much as anything this means that you are _writing_ UTF8 strings which means someone using e.g. C++ API has to read them as such which will cause problems down the line. You got a moment to make a stub code which makes e.g. an HDF5 file containing three strings which illustrates this problem? Will look to help there 🙂

Would have a go myself but load balancing & also may not help you unless we start from how you are making the files.

Reply:
Anyone using C++ already has unicode problems, so writing UTF8 should be well supported by any readers.  

That said, I did a bit more digging.  If I use numpy.string_ everywhere where I use strings in my file creation code, then I can actually eliminate UTF8.  For example here:

https://github.com/cctbx/cctbx_project/blob/master/xfel/swissfel/jf16m_cxigeom2nexus.py#L98

But that can't be the right solution. I think the reader should be able handle unicode or byte strings, since they are both part of the specification.

Reply:
just wanted to point out that I encountered `numpy.string_()` in the past (#56) and there is a link in that pull request with further, possibly relevant information.